### PR TITLE
Revamping Kafka Streams docs for functional style

### LINKED ISSUE #703: Document functional Kafka Streams programming model
- [ ] Improvements to SerDes (new: `CollectionSerde`)
- [ ] Programming model updates
- [ ]

### PR DESCRIPTION
**Pull request by sobychacko**
Resolves #703

### CODE REVIEW

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 1]
```diff
@@ -322,191 +468,343 @@ public static class WordCountProcessorApplication {
...
+spring.cloud.stream.kafka.streams.binder.configuration.default.key.serde=org.apache.kafka.common.serialization.Serdes$StringSerde
+spring.cloud.stream.kafka.streams.binder.configuration.default.value.serde=org.apache.kafka.common.serialization.Serdes$StringSerde
+spring.cloud.stream.kafka.streams.binder.configuration.commit.interval.ms=1000
 ----
 
-If this property is not set, it will use the default SerDe: `spring.cloud.stream.kafka.streams.binder.configuration.default.value.serde`.
+For more information about all the properties that may go into streams configuration, see StreamsConfig JavaDocs in
+Apache Kafka Streams docs.
+
+brokers::
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: Does this section require formatting with code blocks?

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 2]
```diff
@@ -14,61 +14,113 @@ Maven coordinates:
...
+This is a consumer application with no outbound binding and only a single inbound binding.
+The application consumes data and do some terminal operations on that data.
+The application contains the `SpringBootApplication` annotation and a method that is marked as `Bean`.
+The bean method is of type java.util.function.Consumer which is parameterized with `KStream`.
+Then in the implementation, we are returning a Consumer object that is essentially a lambda expression.
+Inside the lambda expression, the code for processing the data is provided.
+
+Kafka Streams binder for Spring Cloud Stream automatically creates the binding behind the scenes.
+In this application, there is a single input binding that is of type KStream.
+The binder creates this binding for the application with a name `process_in`, i.e. the name of the function bean name followed by an underscore and the literal `in`.
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply docs/src/main/asciidoc/kafka-streams.adoc]: not clear what you mean in this particular comment.

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 3]
```diff
@@ -262,26 +372,62 @@ interface KStreamKTableBinding extends KafkaStreamsProcessor {
...
+            @Input("input-3")
+            GlobalKTable<?, ?> input3();
+
+            @Output("output")
+            KStream<?, ?> output();
+    }
+
+----
+
+You might notices that the above two examples are even more verbose since in addition to provide `EnableBinding`, you also need to write your own custom binding interface as well.
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply docs/src/main/asciidoc/kafka-streams.adoc]: can you pls elaborate by what you mean by NOTE?

File: docs/src/main/asciidoc/kafka-streams.adoc (hunk 1 above)
Comment (sobychacko) [reply docs/src/main/asciidoc/kafka-streams.adoc]: hmm - not sure. I didn't change this section much from what was there previously.

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 4]
```diff
@@ -14,61 +14,110 @@ Maven coordinates:
...
 
-Spring Cloud Stream's Apache Kafka support also includes a binder implementation designed explicitly for Apache Kafka
-Streams binding. With this native integration, a Spring Cloud Stream "processor" application can directly use the
+Spring Cloud Stream includes a binder implementation designed explicitly for https://kafka.apache.org/documentation/streams/[Apache Kafka Streams] binding.
+With this native integration, a Spring Cloud Stream "processor" application can directly use the
 https://kafka.apache.org/documentation/streams/developer-guide[Apache Kafka Streams] APIs in the core business logic.
 
-Kafka Streams binder implementation builds on the foundation provided by the https://docs.spring.io/spring-kafka/reference/html/_reference.html#kafka-streams[Kafka Streams in Spring Kafka]
-project.
+Kafka Streams binder implementation builds on the foundations provided by the https://docs.spring.io/spring-kafka/reference/html/#kafka-streams[Spring for Apache Kafka] project
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: Missing `.` in the end

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 5]
```diff
@@ -14,61 +14,110 @@ Maven coordinates:
...
+
+Albeit simple, this is a complete standalone Spring Boot application that is leveraging Kafka Streams for stream processing.
+This is a consumer application with no outbound binding and only a single inbound binding.
+The application consumes data and it simply logs the transformation as standard output.
+The application contains the `SpringBootApplication` annotation and a method that is marked as `Bean`.
+The bean method is of type `java.util.function.Consumer` which is parameterized with `KStream`.
+Then in the implementation, we are returning a Consumer object that is essentially a lambda expression.
+Inside the lambda expression, the code for processing the data is provided.
+
+In this application, there is a single input binding that is of type KStream.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: `KStream`

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 6]
```diff
@@ -79,126 +128,208 @@ KStream objects. As a developer, you can exclusively focus on the business aspec
...
+                )
+    );
+}
+----
 
-==== TimeWindow properties:
+In this model, we have 3 partial functions as inputs. The first function has the first input binding of the application (`Order`) and its output is another function.
+This output function's input is the second input binding for the application (`Customer`) and its output is another function.
+This output function's input is the third input for the application (Product) and its output is a KStream which is final output binding for the application.
+The input from the three partial functions which are KStream, GlobalKTable, GlobalKTable respectively are available for you in the method body for implementing the business logic as part of the lambda expression.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: We could backtick objects whenever we use them, for consistency.

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 7]
```diff
@@ -79,126 +128,208 @@ KStream objects. As a developer, you can exclusively focus on the business aspec
...
+            .map((key, value) -> new KeyValue<>(null, new WordCount(key.key(), value,
+                    new Date(key.window().start()), new Date(key.window().end()))))
+            .branch(isEnglish, isFrench, isSpanish);
+}
+----
+
+The programming model remains the same, however the outbound parameterized type is `KStream[]`.
+The default output binding names are `process_out_0`, `process_out_1`, `process_out_2` respectively.
+
+===== Summary of various functional styles
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: > Function Programming Styles for Kafka Streams

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 8]
```diff
@@ -79,126 +128,208 @@ KStream objects. As a developer, you can exclusively focus on the business aspec
...
+
+|1|0|java.util.function.Consumer
+|2|0|java.util.function.BiConsumer
+|1|1..n (*)|java.util.function.Function
+|2|1..n (*)|java.util.function.BiFunction
+|>= 3 |0..n (*) |Use curried functions
+
+|===
+
+*In the case of more then one output, the type simply becomes `KStream[]`.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: > * In the case of more than one output, the type simply becomes `KStream[]`.

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 9]
```diff
@@ -322,191 +472,345 @@ public static class WordCountProcessorApplication {
 }
 ----
 
-Properties:
+In this secion, we saw the various ways the Kafka Streams binder enables the applications to be written in vaious programming styles.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: > To recap, we have reviewed the various programming model choices when using the Kafka Streams binder.

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 10]
```diff
@@ -322,191 +472,345 @@ public static class WordCountProcessorApplication {
...
-
-===== Outbound serialization
-
-If native encoding is disabled (which is the default), then the framework will convert the message using the contentType
-set by the user (otherwise, the default `application/json` will be applied). It will ignore any SerDe set on the outbound
-in this case for outbound serialization.
-
-Here is the property to set the contentType on the outbound.
+The binder provides bining capabilities for `KStream`, `KTable` and `GlobalKTable` on the input.
+KTable and GlobalKTable bindings are only available on the input.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: backticks

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 11]
```diff
@@ -322,191 +468,343 @@ public static class WordCountProcessorApplication {
...
+
+If the application does not provide an application ID, then in that case the binder will auto generate a random application ID for you.
+This is convenient in development scenarios as it avoids the need for explicitly providing the application ID.
+Please keep in mind that when you rely on this, each time you start the application, it starts with a brand new application id.
+In the case of functional model, the generated application ID will be the function bean name followed by a `UUID` which is then postfixed with the literal `applicationID`.
+In the case `StreamListener`, instead of using the function bean name, the generated application ID will be use the containing class name followed by the method name.
+
+Here is the order of precedence for application ID in a nutshell (from lowest to highest)
+
+* Auto generated by the binder
```
[the remark this answers is not in the document]
Comment (sabbyanandan) [reply docs/src/main/asciidoc/kafka-streams.adoc]: The nested parens is making it hard to read. Can we put this in a table with columns, so it is easy on the eyes?

File: docs/src/main/asciidoc/kafka-streams.adoc [hunk 12]
```diff
@@ -322,191 +472,345 @@ public static class WordCountProcessorApplication {
...
+In the case of `StreamListener`, instead of using the function bean name, the generated application ID will be use the containing class name followed by the method name.
+
+Here is the order of precedence for application ID in a nutshell (from lowest to highest)
+
+* Auto generated by the binder
+* spring.kafka.streams.applicationId or spring.application.name (Can only be used if you have a single processor in the application)
+* binder level (spring.cloud.stream.binder.kafka.streams.applicationId) (only use if you have single processor)
+* binding level (spring.cloud.stream.kafka.streams.bindings.process_in.applicationId _change process_in accordingly_.)
+
+==== Overriding default input/output binding names generated by the binder when using the functional style.
```
Comment (sabbyanandan) [inline docs/src/main/asciidoc/kafka-streams.adoc]: I still think this is a very long sub-title.

File: docs/src/main/asciidoc/kafka-streams.adoc (hunk 1 above)
Comment (sabbyanandan) [reply docs/src/main/asciidoc/kafka-streams.adoc]: No, I was wrong. I see them rendering correctly in the generated docs.

**Comment by sobychacko** [comment]
@sabbyanandan Thanks for the review. I pushed another commit addressing most of your review.

**Pull request closed without merging**